# Состояние онко помощи в РФ в 2021г.
**Задача: Визуализировать ежегодный отчет главного онколога РФ за 2021 год.**

Даны 32 таблицы, где во всех строках регионы РФ  в каждой новой таблице показатели качества организации онкологической помощи для разных регионов. Фактически это основной источник рейтинга онкологического благополучия регионов России. 
Таблицы с 25 по 51 абсолютно идентичны по формату, но содержат данные либо для различных возрастных групп, либо (в подавляющем большинстве) для ЗНО различных локализаций. Важно, что эти 25 таблиц (с 25 по 51) все расположены на двух листах (вкладках) Excel.

Таблица 22 - сведение о больных с ЗНО, не состоявших на учете и умерших в 2021г.  
Таблица 24 (1й лист), 24пр (2й лист) - сведения о контингенте больных с ЗНО, состоящем на учете в 2021г всего (С00-96) 

In [1]:
import pandas as pd
import numpy as np

In [2]:
t_22=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_022_Состояние_онко_помощи_в_РФ.xlsx')

In [3]:
# загрузим первые листы таблиц.
t_24=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_024_Состояние_онко_помощи_в_РФ.xlsx')
t_25=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_025_Состояние_онко_помощи_в_РФ.xlsx')
t_26=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_026_Состояние_онко_помощи_в_РФ.xlsx')
t_27=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_027_Состояние_онко_помощи_в_РФ.xlsx')
t_28=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_028_Состояние_онко_помощи_в_РФ.xlsx')
t_29=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_029_Состояние_онко_помощи_в_РФ.xlsx')
t_30=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_030_Состояние_онко_помощи_в_РФ.xlsx')
t_31=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_031_Состояние_онко_помощи_в_РФ.xlsx')
t_32=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_032_Состояние_онко_помощи_в_РФ.xlsx')
t_33=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_033_Состояние_онко_помощи_в_РФ.xlsx')
t_34=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_034_Состояние_онко_помощи_в_РФ.xlsx')
t_35=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_035_Состояние_онко_помощи_в_РФ.xlsx')
t_36=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_036_Состояние_онко_помощи_в_РФ.xlsx')
t_37=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_037_Состояние_онко_помощи_в_РФ.xlsx')
t_38=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_038_Состояние_онко_помощи_в_РФ.xlsx')
t_39=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_039_Состояние_онко_помощи_в_РФ.xlsx')
t_40=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_040_Состояние_онко_помощи_в_РФ.xlsx')
t_41=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_041_Состояние_онко_помощи_в_РФ.xlsx')
t_42=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_042_Состояние_онко_помощи_в_РФ.xlsx')
t_43=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_043_Состояние_онко_помощи_в_РФ.xlsx')
t_44=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_044_Состояние_онко_помощи_в_РФ.xlsx')
t_45=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_045_Состояние_онко_помощи_в_РФ.xlsx')
t_46=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_046_Состояние_онко_помощи_в_РФ.xlsx')
t_47=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_047_Состояние_онко_помощи_в_РФ.xlsx')
t_48=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_048_Состояние_онко_помощи_в_РФ.xlsx')
t_49=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_049_Состояние_онко_помощи_в_РФ.xlsx')
t_50=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_050_Состояние_онко_помощи_в_РФ.xlsx')
t_51=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_051_Состояние_онко_помощи_в_РФ.xlsx')
t_52=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_052_Состояние_онко_помощи_в_РФ.xlsx')
t_53=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_053_Состояние_онко_помощи_в_РФ.xlsx')

In [4]:
# загрузим вторые листы всех таблиц.
t_24_sheet2=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_024_пр_Состояние_онко_помощи_в_РФ.xlsx')
t_25_sheet2=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_025_Состояние_онко_помощи_в_РФ.xlsx',sheet_name='Table 3')
t_26_sheet2=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_026_Состояние_онко_помощи_в_РФ.xlsx',sheet_name='Table 3')
t_27_sheet2=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_027_Состояние_онко_помощи_в_РФ.xlsx',sheet_name='Table 3')
t_28_sheet2=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_028_Состояние_онко_помощи_в_РФ.xlsx',sheet_name='Table 3')
t_29_sheet2=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_029_Состояние_онко_помощи_в_РФ.xlsx',sheet_name='Table 3')
t_30_sheet2=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_030_Состояние_онко_помощи_в_РФ.xlsx',sheet_name='Table 3')
t_31_sheet2=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_031_Состояние_онко_помощи_в_РФ.xlsx',sheet_name='Table 3')
t_32_sheet2=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_032_Состояние_онко_помощи_в_РФ.xlsx',sheet_name='Table 3')
t_33_sheet2=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_033_Состояние_онко_помощи_в_РФ.xlsx',sheet_name='Table 3')
t_34_sheet2=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_034_Состояние_онко_помощи_в_РФ.xlsx',sheet_name='Table 3')
t_35_sheet2=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_035_Состояние_онко_помощи_в_РФ.xlsx',sheet_name='Table 3')
t_36_sheet2=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_036_Состояние_онко_помощи_в_РФ.xlsx',sheet_name='Table 3')
t_37_sheet2=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_037_Состояние_онко_помощи_в_РФ.xlsx',sheet_name='Table 3')
t_38_sheet2=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_038_Состояние_онко_помощи_в_РФ.xlsx',sheet_name='Table 3')
t_39_sheet2=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_039_Состояние_онко_помощи_в_РФ.xlsx',sheet_name='Table 3')
t_40_sheet2=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_040_Состояние_онко_помощи_в_РФ.xlsx',sheet_name='Table 3')
t_41_sheet2=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_041_Состояние_онко_помощи_в_РФ.xlsx',sheet_name='Table 3')
t_42_sheet2=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_042_Состояние_онко_помощи_в_РФ.xlsx',sheet_name='Table 3')
t_43_sheet2=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_043_Состояние_онко_помощи_в_РФ.xlsx',sheet_name='Table 3')
t_44_sheet2=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_044_Состояние_онко_помощи_в_РФ.xlsx',sheet_name='Table 3')
t_45_sheet2=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_045_Состояние_онко_помощи_в_РФ.xlsx',sheet_name='Table 3')
t_46_sheet2=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_046_Состояние_онко_помощи_в_РФ.xlsx',sheet_name='Table 3')
t_47_sheet2=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_047_Состояние_онко_помощи_в_РФ.xlsx',sheet_name='Table 3')
t_48_sheet2=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_048_Состояние_онко_помощи_в_РФ.xlsx',sheet_name='Table 3')
t_49_sheet2=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_049_Состояние_онко_помощи_в_РФ.xlsx',sheet_name='Table 3')
t_50_sheet2=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_050_Состояние_онко_помощи_в_РФ.xlsx',sheet_name='Table 3')
t_51_sheet2=pd.read_excel('C:/Users/78585/Desktop/Медицинский дашборд/2021_Состояние_онко_помощи_в_РФ_Таблицы_022_и_024-053/2021_Таблица_051_Состояние_онко_помощи_в_РФ.xlsx',sheet_name='Table 5')

In [5]:
full_list_of_tables = [t_22, t_24, t_25, t_26, t_27, t_28, t_29, t_30, t_31, t_32, t_33, t_34, t_35, t_36, t_37, t_38, t_39,\
                       t_40, t_41, t_42, t_43, t_44, t_45, t_46, t_47, t_48, t_49, t_50, t_51, t_52, t_53, t_24_sheet2,\
                       t_25_sheet2, t_26_sheet2, t_27_sheet2, t_28_sheet2, t_29_sheet2, t_30_sheet2, t_31_sheet2,\
                       t_32_sheet2, t_33_sheet2, t_34_sheet2, t_35_sheet2, t_36_sheet2, t_37_sheet2, t_38_sheet2,\
                       t_39_sheet2, t_40_sheet2, t_41_sheet2, t_42_sheet2, t_43_sheet2, t_44_sheet2, t_45_sheet2,\
                       t_46_sheet2, t_47_sheet2, t_48_sheet2, t_49_sheet2, t_50_sheet2, t_51_sheet2]
first_sheet_of_tables = [t_24, t_25, t_26, t_27, t_28, t_29, t_30, t_31, t_32, t_33, t_34, t_35, t_36, t_37, t_38, t_39,\
                         t_40, t_41, t_42, t_43, t_44, t_45, t_46, t_47, t_48, t_49, t_50, t_51, t_52, t_53]
second_sheet_of_tables = [t_24_sheet2,t_25_sheet2, t_26_sheet2, t_27_sheet2, t_28_sheet2, t_29_sheet2, t_30_sheet2,\
                          t_31_sheet2,t_32_sheet2, t_33_sheet2, t_34_sheet2, t_35_sheet2, t_36_sheet2, t_37_sheet2,\
                          t_38_sheet2, t_39_sheet2, t_40_sheet2, t_41_sheet2, t_42_sheet2, t_43_sheet2, t_44_sheet2,\
                          t_45_sheet2,t_46_sheet2, t_47_sheet2, t_48_sheet2, t_49_sheet2, t_50_sheet2, t_51_sheet2]

## 1. Переименование столбцов.

In [6]:
# переименуем столбцы столбцы в таблице 22
t_22.columns = ['Республика, край, область',\
                'Умершие, не состаявшие на учете, абсолютное число',\
                'Умершие, не состаявшие на учете, на 100 умерших от ЗНО',\
                'Умершие, не состаявшие на учете, на 100 больных с впервые в жизни установленным диагнозом',\
                'Диагноз установлен посмертно, абсолютное число',\
                'Диагноз установлен посмертно на 100 больных с впервые в жизни установленным диагнозом',\
                'Диагноз установлен при вскрытии, абсолютное число',\
                'Диагноз установлен при вскрытии на 100 больных с впервые в жизни установленным диагнозом',\
                'Диагноз установлен при вскрытии, % от всех установленных посмертно']

In [33]:
t_22.head()

Республика, край, область Умершие, не состаявшие на учете, абсолютное число  \
0      Белгородская область                                               112   
1          Брянская область                                               236   
2      Владимирская область                                               598   
3       Воронежская область                                               172   
4        Ивановская область                                               199   

  Умершие, не состаявшие на учете, на 100 умерших от ЗНО  \
0                                                4.6       
1                                                8.8       
2                                               23.7       
3                                                4.3       
4                                               10.5       

  Умершие, не состаявшие на учете, на 100 больных с впервые в жизни установленным диагнозом  \
0                                                1.8                                          
1                                                4.6                                          
2                                                 12                                          
3                                                2.1                                          
4                                                4.8                                          

  Диагноз установлен посмертно, абсолютное число  \
0                                            112   
1                                            234   
2                                            589   
3                                            172   
4                                            199   

  Диагноз установлен посмертно на 100 больных с впервые в жизни установленным диагнозом  \
0                                                1.8                                      
1                                                4.6                                      
2                                               11.8                                      
3                                                2.1                                      
4                                                4.8                                      

  Диагноз установлен при вскрытии, абсолютное число  \
0                                               112   
1                                               234   
2                                               585   
3                                               172   
4                                               199   

  Диагноз установлен при вскрытии на 100 больных с впервые в жизни установленным диагнозом  \
0                                                1.8                                         
1                                                4.6                                         
2                                               11.7                                         
3                                                2.1                                         
4                                                4.8                                         

  Диагноз установлен при вскрытии, % от всех установленных посмертно  \
0                                                100                   
1                                                100                   
2                                               99.3                   
3                                                100                   
4                                                100                   

  Федеральный округ  
0       Центральный  
1       Центральный  
2       Центральный  
3       Центральный  
4       Центральный

In [8]:
# создадим функцию для переименования столбцов в таблицах на первом листе
def new_columns_sheet1 (df):
    df.columns = ['Республика, край, область',\
                  'Взято на учет с впервые в жизни установленным диагнозом, абсолютное число',\
                  'Диагноз выявлен активно, %',\
                  'На учете на конец года, абсолютное число',\
                  'На учете на конец года, на 100 тыс.населения',\
                  'На учете на конец года, 5 лет и более, абсолютное число',\
                  'На учете на конец года, 5 лет и более, % от состоящих на учете',\
                  'Индекс накопления контингентов',\
                  'Летальность,%']

for df in first_sheet_of_tables:
    new_columns_sheet1(df)

In [34]:
t_35.head()

Республика, край, область  \
0  Республика, край, область   
1                        NaN   
2                     РОССИЯ   
3             ЦЕНТРАЛЬНЫЙ ФО   
4       Белгородская область   

  Взято на учет с впервые в жизни установленным диагнозом, абсолютное число  \
0  Взято на учет больных с впервые в жизни уст. д...                          
1                                                NaN                          
2                                              13845                          
3                                               3605                          
4                                                152                          

      Диагноз выявлен активно, % На учете на конец года, абсолютное число  \
0  в т.ч. вы- явлены активно,\n%        Находились на учете на конец года   
1                            NaN                       абсолют- ное число   
2                            5.8                                    20630   
3                            5.9                                     6479   
4                            3.9                                      168   

  На учете на конец года, на 100 тыс.населения  \
0                                          NaN   
1                        на 100 тыс. населения   
2                                         14.1   
3                                         16.5   
4                                         10.9   

  На учете на конец года, 5 лет и более, абсолютное число  \
0                               из них 5 лет и более        
1                                 абсолют- ное число        
2                                               6951        
3                                               2070        
4                                                 44        

  На учете на конец года, 5 лет и более, % от состоящих на учете  \
0                                                NaN               
1                               %  от сост. на учете               
2                                               33.7               
3                                               31.9               
4                                               26.2               

       Индекс накопления контингентов      Летальность,% Федеральный округ  \
0  Индекс на- копления контин- гентов  Леталь- ность,\n%              None   
1                                 NaN                NaN              None   
2                                 1.5               38.1              None   
3                                 1.8               32.6              None   
4                                 1.1               41.1       Центральный   

                 Локализация  
0  Поджелудочная железа(C25)  
1  Поджелудочная железа(C25)  
2  Поджелудочная железа(C25)  
3  Поджелудочная железа(C25)  
4  Поджелудочная железа(C25)

In [10]:
# создадим функцию для переименования столбцов в таблицах на втором листе
def new_columns_sheet2 (df):
    df.columns = ['Республика, край, область',\
                      'Зарегитрировано ЗНО без учтенных посмертно, абсолютное число',\
                      'Диагноз подтвержден морфологически, %',\
                      'I стадия, %',\
                      'II стадия, %',\
                      'III стадия, %',\
                      'IV стадия, %',\
                      'Стадия не установлена, %',\
                      'Летальность на первом году жизни с момента установления диагноза, %']

for df in second_sheet_of_tables:
    new_columns_sheet2(df)

## 2. Создание столбца "Федеральный округ".

In [11]:
# создадим списки Федеральных округов с относящимися к ним территориям.
central_fo = list(t_35.loc[4:21,'Республика, край, область'])
nord_west_fo = list(t_35.loc[23:33, 'Республика, край, область'])
south_fo = list(t_35.loc[35:42,'Республика, край, область'])
nord_kavkaz_fo = list(t_35.loc[44:50, 'Республика, край, область'])
privolzh_fo = list(t_35.loc[52:65,'Республика, край, область'])
ural_fo = list(t_35.loc[67:72, 'Республика, край, область'])
siberia_fo = list(t_35.loc[74:83,'Республика, край, область'])
dalnevostok_fo = list(t_35.loc[85:95, 'Республика, край, область'])

siberia_fo

['Алтайский край',
 'Красноярский край',
 'Иркутская область',
 'Кемеровская область',
 'Новосибирская область',
 'Омская область',
 'Томская область',
 'Республика Алтай',
 'Республика Тыва',
 'Республика Хакасия']

In [12]:
# функция для заполнения столбца с Федеральным округом.
def fo (region):
    if region in central_fo:
        return 'Центральный'
    if region in nord_west_fo:
        return 'Северо-Западный'
    if region in south_fo:
        return 'Южный'
    if region in nord_kavkaz_fo:
        return 'Северо-Кавказский'
    if region in privolzh_fo:
        return 'Приволжский'
    if region in ural_fo:
        return 'Уральский'
    if region in siberia_fo:
        return 'Сибиркий'
    if region in dalnevostok_fo:
        return 'Северо-Западный'    

In [13]:
# с помощью цикла создадим новый столбец во всех таблицах и заполним его соответствующим Федеральным округом.
for df in full_list_of_tables:
    df['Федеральный округ'] = df['Республика, край, область'].apply(fo)

In [35]:
t_44_sheet2.head()

Республика, край, область  \
0                       NaN   
1                       NaN   
2                       NaN   
3                    РОССИЯ   
4            ЦЕНТРАЛЬНЫЙ ФО   

  Зарегитрировано ЗНО без учтенных посмертно, абсолютное число  \
0    Зарегистриро- вано ЗНО (без учтенных посмертно)             
1                                                NaN             
2                                                NaN             
3                                              24705             
4                                               7462             

     Диагноз подтвержден морфологически, %                  I стадия, %  \
0                                   из них                          NaN   
1  диагноз подтвержден морфологи- чески, %  имели стадию заболевания, %   
2                                      NaN                            I   
3                                       99                         70.8   
4                                     99.2                         73.2   

  II стадия, % III стадия, % IV стадия, % Стадия не установлена, %  \
0          NaN           NaN          NaN                      NaN   
1          NaN           NaN          NaN                      NaN   
2           II           III           IV         не уста- новлена   
3         13.6           8.9          5.9                      0.8   
4         12.4           7.9          5.3                      1.2   

  Летальность на первом году жизни с момента установления диагноза, %  \
0  Летальность на первом году с момента уст. диаг...                    
1                                                NaN                    
2                                                NaN                    
3                                                  7                    
4                                                6.2                    

  Федеральный округ  
0              None  
1              None  
2              None  
3              None  
4              None

 ## 3. Создание столбца с локализацией ЗНО.

In [15]:
t_27['Локализация'] = 'Губа(C00)'
t_28['Локализация'] = 'Полость рта(C01-09)'
t_29['Локализация'] = 'Глотка(C10-13)'
t_30['Локализация'] = 'Пищевод(C15)'
t_31['Локализация'] = 'Желудок(C16)'
t_32['Локализация'] = 'Ободочная кишка(С18)'
t_33['Локализация'] = 'Прямая кишка, ректосиг. соединение, анус(19-21)'
t_34['Локализация'] = 'Печень и желчные протоки(С22)'
t_35['Локализация'] = 'Поджелудочная железа(C25)'
t_36['Локализация'] = 'Гортань(C32)'
t_37['Локализация'] = 'Трахея, бронхи, легкие(С33,34)'
t_38['Локализация'] = 'Кости и суставные хрящи(C40,41)'
t_39['Локализация'] = 'Меланома кожи(С43)'
t_40['Локализация'] = 'Кожа(кроме меланомы)(С44)'
t_41['Локализация'] = 'Соединительные и другие мягкие ткани(С47,49)'
t_42['Локализация'] = 'Молочная железа(50)'
t_43['Локализация'] = 'Шейка матки(C53)'
t_44['Локализация'] = 'Тело матки(C54)'
t_45['Локализация'] = 'Яичник(C56)'
t_46['Локализация'] = 'Предстательная железа(C61)'
t_47['Локализация'] = 'Почка(C64)'
t_48['Локализация'] = 'Мочевой пузырь(C67)'
t_49['Локализация'] = 'Щитовидная железа(C73)'
t_50['Локализация'] = 'Злокачественные лимфомы(С91-86,88,90,96)'
t_51['Локализация'] = 'Лейкемии(C91-95)'

In [36]:
t_43.head()

Республика, край, область  \
0  Республика, край, область   
1                        NaN   
2                     РОССИЯ   
3             ЦЕНТРАЛЬНЫЙ ФО   
4       Белгородская область   

  Взято на учет с впервые в жизни установленным диагнозом, абсолютное число  \
0  Взято на учет больных с впервые в жизни уст. д...                          
1                                                NaN                          
2                                              14468                          
3                                               3303                          
4                                                129                          

      Диагноз выявлен активно, % На учете на конец года, абсолютное число  \
0  в т.ч. вы- явлены активно,\n%        Находились на учете на конец года   
1                            NaN                       абсолют- ное число   
2                           34.7                                   185507   
3                           34.8                                    48430   
4                           48.8                                     1940   

  На учете на конец года, на 100 тыс.населения  \
0                                          NaN   
1                        на 100 тыс. населения   
2                                        126.7   
3                                        123.1   
4                                        125.5   

  На учете на конец года, 5 лет и более, абсолютное число  \
0                               из них 5 лет и более        
1                                 абсолют- ное число        
2                                             125247        
3                                              32973        
4                                               1361        

  На учете на конец года, 5 лет и более, % от состоящих на учете  \
0                                                NaN               
1                               %  от сост. на учете               
2                                               67.5               
3                                               68.1               
4                                               70.2               

       Индекс накопления контингентов      Летальность,% Федеральный округ  \
0  Индекс на- копления контин- гентов  Леталь- ность,\n%              None   
1                                 NaN                NaN              None   
2                                12.8                2.9              None   
3                                14.7                2.5              None   
4                                  15                2.3       Центральный   

        Локализация  
0  Шейка матки(C53)  
1  Шейка матки(C53)  
2  Шейка матки(C53)  
3  Шейка матки(C53)  
4  Шейка матки(C53)

## 4. Объединение двух листов таблиц.

In [17]:
t24=t_24.merge(t_24_sheet2, on='Республика, край, область', how='outer')
t25=t_25.merge(t_25_sheet2, on='Республика, край, область', how='outer')
t26=t_26.merge(t_26_sheet2, on='Республика, край, область', how='outer')
t27=t_27.merge(t_27_sheet2, on='Республика, край, область', how='outer')
t28=t_28.merge(t_28_sheet2, on='Республика, край, область', how='outer')
t29=t_29.merge(t_29_sheet2, on='Республика, край, область', how='outer')
t30=t_30.merge(t_30_sheet2, on='Республика, край, область', how='outer')
t31=t_31.merge(t_31_sheet2, on='Республика, край, область', how='outer')
t32=t_32.merge(t_32_sheet2, on='Республика, край, область', how='outer')
t33=t_33.merge(t_33_sheet2, on='Республика, край, область', how='outer')
t34=t_34.merge(t_34_sheet2, on='Республика, край, область', how='outer')
t35=t_35.merge(t_35_sheet2, on='Республика, край, область', how='outer')
t36=t_36.merge(t_36_sheet2, on='Республика, край, область', how='outer')
t37=t_37.merge(t_37_sheet2, on='Республика, край, область', how='outer')
t38=t_38.merge(t_38_sheet2, on='Республика, край, область', how='outer')
t39=t_39.merge(t_39_sheet2, on='Республика, край, область', how='outer')
t40=t_40.merge(t_40_sheet2, on='Республика, край, область', how='outer')
t41=t_41.merge(t_41_sheet2, on='Республика, край, область', how='outer')
t42=t_42.merge(t_42_sheet2, on='Республика, край, область', how='outer')
t43=t_43.merge(t_43_sheet2, on='Республика, край, область', how='outer')
t44=t_44.merge(t_44_sheet2, on='Республика, край, область', how='outer')
t45=t_45.merge(t_45_sheet2, on='Республика, край, область', how='outer')
t46=t_46.merge(t_46_sheet2, on='Республика, край, область', how='outer')
t47=t_47.merge(t_47_sheet2, on='Республика, край, область', how='outer')
t48=t_48.merge(t_48_sheet2, on='Республика, край, область', how='outer')
t49=t_49.merge(t_49_sheet2, on='Республика, край, область', how='outer')
t50=t_50.merge(t_50_sheet2, on='Республика, край, область', how='outer')
t51=t_51.merge(t_51_sheet2, on='Республика, край, область', how='outer')

In [18]:
# создадим список таблиц с локализацией ЗНО.
double_table = [t27, t28, t29, t30, t31, t32, t33, t34, t35, t36, t37, t38, t39, t40, t41, t42, t43, t44, t45, t46, t47,\
    t48, t49, t50, t51,]

In [19]:
for df in double_table:
    print(df.shape)

(97, 20)
(97, 20)
(97, 20)
(97, 20)
(97, 20)
(98, 20)
(97, 20)
(98, 20)
(98, 20)
(98, 20)
(97, 20)
(97, 20)
(97, 20)
(97, 20)
(98, 20)
(97, 20)
(97, 20)
(98, 20)
(97, 20)
(97, 20)
(97, 20)
(97, 20)
(97, 20)
(97, 20)
(97, 20)


Увеличение числа строк в некоторых таблицах связано с тем, что при конвертаци таблиц из формата PDF в Excel в таблицах 32,34,41 номер страницы был распознан как отдельная строка. А в таблицах 35,36 нет наименования столбца с регионами, и при соединении 2 листов методом merge образуется дополнительная строка. В дальнейшем все эти строки будут удалены.

In [20]:
t41.tail(2)

Республика, край, область  \
96       Еврейская авт. обл.   
97                       111   

   Взято на учет с впервые в жизни установленным диагнозом, абсолютное число  \
96                                                  3                          
97                                                NaN                          

   Диагноз выявлен активно, % На учете на конец года, абсолютное число  \
96                          0                                       33   
97                        NaN                                      NaN   

   На учете на конец года, на 100 тыс.населения  \
96                                           21   
97                                          NaN   

   На учете на конец года, 5 лет и более, абсолютное число  \
96                                                 24        
97                                                NaN        

   На учете на конец года, 5 лет и более, % от состоящих на учете  \
96                                               72.7               
97                                                NaN               

   Индекс накопления контингентов Летальность,% Федеральный округ_x  \
96                             11             0     Северо-Западный   
97                            NaN           NaN                 NaN   

                                     Локализация  \
96  Соединительные и другие мягкие ткани(С47,49)   
97                                           NaN   

   Зарегитрировано ЗНО без учтенных посмертно, абсолютное число  \
96                                                  5             
97                                                NaN             

   Диагноз подтвержден морфологически, % I стадия, % II стадия, %  \
96                                   100          40           20   
97                                   NaN         NaN          NaN   

   III стадия, % IV стадия, % Стадия не установлена, %  \
96            40            0                        0   
97           NaN          NaN                      NaN   

   Летальность на первом году жизни с момента установления диагноза, %  \
96                                                  0                    
97                                                NaN                    

   Федеральный округ_y  
96     Северо-Западный  
97                None

## 5. Соединение всех таблиц с локализацией в одну.

In [37]:
big_table = pd.concat(double_table, axis=0).reset_index(drop=True)
big_table.head()

Республика, край, область  \
0  Республика, край, область   
1                        NaN   
2                        NaN   
3                     РОССИЯ   
4             ЦЕНТРАЛЬНЫЙ ФО   

  Взято на учет с впервые в жизни установленным диагнозом, абсолютное число  \
0  Взято на учет больных с впервые в жизни уст. д...                          
1                                                NaN                          
2                                                NaN                          
3                                               1675                          
4                                                308                          

      Диагноз выявлен активно, % На учете на конец года, абсолютное число  \
0  в т.ч. вы- явлены активно,\n%        Находились на учете на конец года   
1                            NaN                       абсолют- ное число   
2                            NaN                       абсолют- ное число   
3                           35.4                                    32914   
4                           41.6                                     6239   

  На учете на конец года, на 100 тыс.населения  \
0                                          NaN   
1                        на 100 тыс. населения   
2                        на 100 тыс. населения   
3                                         22.5   
4                                         15.9   

  На учете на конец года, 5 лет и более, абсолютное число  \
0                               из них 5 лет и более        
1                                 абсолют- ное число        
2                                 абсолют- ное число        
3                                              24985        
4                                               4679        

  На учете на конец года, 5 лет и более, % от состоящих на учете  \
0                                                NaN               
1                               %  от сост. на учете               
2                               %  от сост. на учете               
3                                               75.9               
4                                                 75               

       Индекс накопления контингентов      Летальность,% Федеральный округ_x  \
0  Индекс на- копления контин- гентов  Леталь- ность,\n%                None   
1                                 NaN                NaN                None   
2                                 NaN                NaN                None   
3                                19.7                0.9                None   
4                                20.3                1.2                None   

  Локализация Зарегитрировано ЗНО без учтенных посмертно, абсолютное число  \
0   Губа(C00)    Зарегистриро- вано ЗНО (без учтенных посмертно)             
1   Губа(C00)                                                NaN             
2   Губа(C00)                                                NaN             
3   Губа(C00)                                               1881             
4   Губа(C00)                                                345             

     Диагноз подтвержден морфологически, %                  I стадия, %  \
0                                   из них                          NaN   
1  диагноз подтвержден морфологи- чески, %  имели стадию заболевания, %   
2                                      NaN                            I   
3                                     99.4                           53   
4                                     98.8                         56.5   

  II стадия, % III стадия, % IV стадия, % Стадия не установлена, %  \
0          NaN           NaN          NaN                      NaN   
1          NaN           NaN          NaN                      NaN   
2           II           III           IV         не уста- новлена   
3         30.5          10.5          5.5                      0.5   
4         27.5           9.3          5.5   

In [22]:
# Переименуем столбец ['Федеральный округ_x'] и удалим второй с округами.
big_table = big_table.rename(columns={'Федеральный округ_x':'Федеральный округ'})
big_table = big_table.drop(columns=['Федеральный округ_y'])

## 6. Удалим строчки с пропусками по столбцу 'Федеральный округ'.

Это позволит оставить в датесете строки только с названиями регионов.

In [23]:
big_table = big_table.dropna(subset = ['Федеральный округ']).reset_index(drop=True)

In [24]:
big_table.shape

(2125, 19)

В РФ 85 регионов, краёв и округом и в данных 25 таблиц с локализациями. Следовательно 2125 строк.

In [25]:
big_table.columns

Index(['Республика, край, область',
       'Взято на учет с впервые в жизни установленным диагнозом, абсолютное число',
       'Диагноз выявлен активно, %',
       'На учете на конец года, абсолютное число',
       'На учете на конец года, на 100 тыс.населения',
       'На учете на конец года, 5 лет и более, абсолютное число',
       'На учете на конец года, 5 лет и более, % от состоящих на учете',
       'Индекс накопления контингентов', 'Летальность,%', 'Федеральный округ',
       'Локализация',
       'Зарегитрировано ЗНО без учтенных посмертно, абсолютное число',
       'Диагноз подтвержден морфологически, %', 'I стадия, %', 'II стадия, %',
       'III стадия, %', 'IV стадия, %', 'Стадия не установлена, %',
       'Летальность на первом году жизни с момента установления диагноза, %'],
      dtype='object')

In [26]:
big_table.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2125 entries, 0 to 2124
Data columns (total 19 columns):
 #   Column                                                                     Non-Null Count  Dtype 
---  ------                                                                     --------------  ----- 
 0   Республика, край, область                                                  2125 non-null   object
 1   Взято на учет с впервые в жизни установленным диагнозом, абсолютное число  2125 non-null   object
 2   Диагноз выявлен активно, %                                                 2125 non-null   object
 3   На учете на конец года, абсолютное число                                   2125 non-null   object
 4   На учете на конец года, на 100 тыс.населения                               2125 non-null   object
 5   На учете на конец года, 5 лет и более, абсолютное число                    2125 non-null   object
 6   На учете на конец года, 5 лет и более, % от состоящих на учете  

Эти пропуски связаны с тем, что в 4 таблицах нет некоторых данных по Чукотскому автономногу округу и Магаданской области.

In [27]:
big_table.loc[big_table['I стадия, %'].isna()]

Республика, край, область  \
251        Чукотский авт.округ   
1268       Магаданская область   
1271       Чукотский авт.округ   
1611       Чукотский авт.округ   

     Взято на учет с впервые в жизни установленным диагнозом, абсолютное число  \
251                                                   0                          
1268                                                  0                          
1271                                                  0                          
1611                                                  0                          

     Диагноз выявлен активно, % На учете на конец года, абсолютное число  \
251                           0                                        1   
1268                          0                                        8   
1271                          0                                        3   
1611                          0                                       14   

     На учете на конец года, на 100 тыс.населения  \
251                                             2   
1268                                          5.7   
1271                                            6   
1611                                         28.1   

     На учете на конец года, 5 лет и более, абсолютное число  \
251                                                   1        
1268                                                  6        
1271                                                  0        
1611                                                 11        

     На учете на конец года, 5 лет и более, % от состоящих на учете  \
251                                                 100               
1268                                                 75               
1271                                                  0               
1611                                               78.6               

     Индекс накопления контингентов Летальность,% Федеральный округ  \
251                               0             0   Северо-Западный   
1268                              0            20   Северо-Западный   
1271                              0             0   Северо-Западный   
1611                              0             0   Северо-Западный   

                                       Локализация  \
251                                 Глотка(C10-13)   
1268  Соединительные и другие мягкие ткани(С47,49)   
1271  Соединительные и другие мягкие ткани(С47,49)   
1611                                   Яичник(C56)   

     Зарегитрировано ЗНО без учтенных посмертно, абсолютное число  \
251                                                 NaN             
1268                                                NaN             
1271                                                NaN             
1611                                                NaN             

     Диагноз подтвержден морфологически, % I стадия, % II стадия, %  \
251                                    NaN         NaN          NaN   
1268                                   NaN         NaN          NaN   
1271                                   NaN         NaN          NaN   
1611                                   NaN         NaN          NaN   

     III стадия, % IV стадия, % Стадия не установлена, %  \
251            NaN          NaN                      NaN   
1268           NaN          NaN                      NaN   
1271           NaN          NaN                      NaN   
1611           NaN          NaN                      NaN   

     Летальность на первом году жизни с момента установления диагноза, %  
251                                                 NaN                   
1268                                                NaN                   
1271                                                NaN                   
1611                                                NaN

## 7. Cоздание столбца с группой населения

In [28]:
t24['Группа'] = 'ЗНО всего(С00-96)'
t25['Группа'] = 'Дети 0-14 лет'
t26['Группа'] = 'Дети 0-17 лет'
t_52['Группа'] = 'У лиц > 65'
t_53['Группа'] = 'Cельские жители'

In [29]:
t24 = t24.rename(columns={'Федеральный округ_x':'Федеральный округ'})
t25 = t25.rename(columns={'Федеральный округ_x':'Федеральный округ'})
t26 = t26.rename(columns={'Федеральный округ_x':'Федеральный округ'})

t24 = t24.drop(columns=['Федеральный округ_y'])
t25 = t25.drop(columns=['Федеральный округ_y'])
t26 = t26.drop(columns=['Федеральный округ_y'])

t_22 = t_22.dropna(subset = ['Федеральный округ']).reset_index(drop=True)

In [30]:
group_table = pd.concat([t24, t25, t26, t_52, t_53], axis=0)
group_table = group_table.dropna(subset = ['Федеральный округ']).reset_index(drop=True)
group_table.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 425 entries, 0 to 424
Data columns (total 19 columns):
 #   Column                                                                     Non-Null Count  Dtype 
---  ------                                                                     --------------  ----- 
 0   Республика, край, область                                                  425 non-null    object
 1   Взято на учет с впервые в жизни установленным диагнозом, абсолютное число  425 non-null    object
 2   Диагноз выявлен активно, %                                                 425 non-null    object
 3   На учете на конец года, абсолютное число                                   425 non-null    object
 4   На учете на конец года, на 100 тыс.населения                               425 non-null    object
 5   На учете на конец года, 5 лет и более, абсолютное число                    425 non-null    object
 6   На учете на конец года, 5 лет и более, % от состоящих на учете    

In [38]:
group_table.head()

Республика, край, область  \
0      Белгородская область   
1          Брянская область   
2      Владимирская область   
3       Воронежская область   
4        Ивановская область   

  Взято на учет с впервые в жизни установленным диагнозом, абсолютное число  \
0                                               6134                          
1                                               5137                          
2                                               4986                          
3                                               8170                          
4                                               4129                          

  Диагноз выявлен активно, % На учете на конец года, абсолютное число  \
0                       27.4                                    44114   
1                       20.3                                    38258   
2                       10.6                                    44686   
3                       29.2                                    71536   
4                       28.9                                    31860   

  На учете на конец года, на 100 тыс.населения  \
0                                       2854.9   
1                                       3221.5   
2                                       3309.4   
3                                       3090.2   
4                                       3211.4   

  На учете на конец года, 5 лет и более, абсолютное число  \
0                                              24836        
1                                              21775        
2                                              26554        
3                                              41212        
4                                              18064        

  На учете на конец года, 5 лет и более, % от состоящих на учете  \
0                                               56.3               
1                                               56.9               
2                                               59.4               
3                                               57.6               
4                                               56.7               

  Индекс накопления контингентов Летальность,% Федеральный округ  \
0                            7.2           5.1       Центральный   
1                            7.4             6       Центральный   
2                              9           4.1       Центральный   
3                            8.8           5.1       Центральный   
4                            7.7           5.1       Центральный   

  Зарегитрировано ЗНО без учтенных посмертно, абсолютное число  \
0                                               6925             
1                                               5702             
2                                               5503             
3                                               9021             
4                                               4587             

  Диагноз подтвержден морфологически, % I стадия, % II стадия, %  \
0                                  94.5        36.2         26.2   
1                                  97.1        31.3         18.6   
2                                  96.8        30.4         23.7   
3                                  94.2        38.4         26.6   
4                                    98        33.6           21   

  III стадия, % IV стадия, % Стадия не установлена, %  \
0          15.3         18.6                      3.7   
1            17         28.6                      4.6   
2            18         23.5                      4.5   
3            15         18.5                      1.5   
4          14.6         23.8                      7.1   

  Летальность на первом году жизни с момента установления диагноза, %  \
0                                               18.7                    
1                                               26.4                    
2                                              

In [32]:
# выгружаем готовые датасеты.
# big_table.to_csv('localisation.csv')     # t27-t51
# group_table.to_csv('group_table.csv')    # t24,25,26,52,53
# t_22.to_csv('bez_ucheta.csv')            # t22